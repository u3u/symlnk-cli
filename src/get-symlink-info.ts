import path from 'node:path'
import { bold } from 'kolorist'

export interface GetSymlinkInfoOptions {
  cwd?: string
  dest?: string
  format?(path: string, type: 'source' | 'target'): string
}

const relative = (to: string) => {
  if (path.isAbsolute(to)) return to

  return path.relative('.', to)
}

export const getSymlinkInfo = (file: string, options: GetSymlinkInfoOptions = {}) => {
  const { cwd = '.', dest = '.', format = bold } = options

  const source = path.isAbsolute(file)
    ? file
    : path.isAbsolute(dest)
    ? path.resolve(cwd, file)
    : path.relative(dest, path.join(cwd, file))

  const target = path.join(dest, path.basename(file))
  const link = `${format(relative(target), 'target')} -> ${format(relative(source), 'source')}`

  return {
    link,
    source,
    target,
  }
}
