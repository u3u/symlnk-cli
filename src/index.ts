export * from './create-symlinks'
export * from './get-symlink-info'
