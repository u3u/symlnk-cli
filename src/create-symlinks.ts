import fs from 'node:fs/promises'
import path from 'node:path'
import { type Options } from 'globby'
import { blue, bold, gray, green, red, underline, yellow } from 'kolorist'
import { getSymlinkInfo } from './get-symlink-info'

export interface CreateSymlinksOptions extends Omit<Options, 'cwd'> {
  clean?: boolean
  cwd?: string
  dest?: string
  dry?: boolean
  force?: boolean
  onConfrim?(): Promise<boolean>
  yes?: boolean
}

export const createSymlinks = async (files: string | string[], options: CreateSymlinksOptions = {}) => {
  const { clean, cwd = '.', dest = '.', dry, force, ignore = [], onConfrim, yes, ...globbyOptions } = options
  const needConfirm = !yes && !dry && onConfrim
  const { globby } = await import('globby')

  const paths = await globby(files, {
    cwd,
    ignore,
    onlyFiles: false,
    ...globbyOptions,
  })

  if (!paths.length) {
    console.log(yellow('ℹ'), 'No matching files found.')

    return
  }

  if (needConfirm) {
    for (const file of paths) {
      const { link, target } = getSymlinkInfo(file, {
        cwd,
        dest,
        format: (path, type) => (type === 'source' ? underline(bold(path)) : bold(path)),
      })

      console.log(clean ? red('D') : green('M'), clean ? bold(target) : link)
    }

    if (!(await onConfrim())) {
      console.log(blue('ℹ'), 'Canceled.')

      return
    }
  }

  return Promise.allSettled(
    paths.map(async (item) => {
      const { link, source, target } = getSymlinkInfo(item, { cwd, dest })

      try {
        if (!dry) {
          if (force || clean) await fs.unlink(target).catch(() => {})

          if (!clean) {
            await fs.mkdir(path.dirname(target), { recursive: true })
            await fs.symlink(source, target)
          }
        }

        console.log(green('✔'), clean ? bold(target) : link)
      } catch {
        console.log(gray(`skip ${link}`))
      }
    })
  )
}
