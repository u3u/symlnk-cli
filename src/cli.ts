#!/usr/bin/env node
import { cac } from 'cac'
import { blue } from 'kolorist'
import prompts from 'prompts'
import { name, version } from '../package.json'
import { createSymlinks } from './create-symlinks'

const cli = cac(name)

cli.version(version)

cli
  .command('[...files]', 'Glob patterns for create symlinks')
  .option('--dest [path]', 'Destination path', { default: '.' })
  .option('--cwd [path]', 'The current working directory in which to search', { default: '.' })
  .option('-i, --ignore [...files]', 'Ignore files and folders')
  .option('-f, --force', 'Force symlink creation')
  .option('--clean', 'Clean matched symlinks')
  .option('--dry, --dry-run', 'Skip creating/clean symlinks (only output matching files)')
  .option('-y, --yes', 'Skip confirmation')
  .action(async (files, options) => {
    const { clean, cwd, dest, dry, force, ignore, yes } = options

    const result = await createSymlinks(files, {
      clean,
      cwd,
      dest,
      dry,
      force,
      ignore,

      onConfrim: async () => {
        const result = await prompts({
          initial: true,
          message: `Are you sure you want to ${clean ? 'clean' : 'create'} these symlinks?`,
          name: 'value',
          type: 'confirm',
        })

        return result.value
      },

      yes,
    })

    if (!result) return

    if (dry) console.log(blue('ℹ'), 'Dry run, no changes will be made to symlinks.')
    else console.log('✨', `${clean ? 'Clean' : 'Create'} symlinks done.`)
  })

cli.help()

cli.parse()
