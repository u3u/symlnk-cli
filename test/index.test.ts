import fs from 'node:fs/promises'
import path from 'node:path'
import { options } from 'kolorist'
import { expect, it, vi } from 'vitest'
import { createSymlinks, type CreateSymlinksOptions } from '../src'

// Disable color, otherwise the snapshot will not match.
options.enabled = false

const cwd = 'test/__fixtures__'
const dest = 'temp/symlinks'

it('should match snapshots', async () => {
  const spy = vi.spyOn(console, 'log')

  const createSymlinksToMatchSnapshot = async (
    files: string | string[],
    options: CreateSymlinksOptions & { message?: string } = {}
  ) => {
    const { message, ...others } = options

    spy.mockClear()
    await createSymlinks(files, others)
    expect(spy.mock.calls.sort().map((item) => item.join(' '))).toMatchSnapshot(message)
  }

  await createSymlinksToMatchSnapshot('*.nginx', {
    message: 'no match',
  })

  await createSymlinksToMatchSnapshot('*', {
    clean: true,
    cwd,
    dest,
    message: 'clean',
  })

  await createSymlinksToMatchSnapshot('*', {
    cwd,
    dest,
    message: 'create',
    onConfrim: async () => true,
  })

  await createSymlinksToMatchSnapshot('*', {
    cwd,
    dest,
    message: 'skip exists',
  })

  await createSymlinksToMatchSnapshot('*', {
    cwd,
    dest: 'test',
    message: 'cancel create',
    onConfrim: async () => false,
  })

  await createSymlinksToMatchSnapshot('*', {
    clean: true,
    cwd,
    dest,
    message: 'cancel clean',
    onConfrim: async () => false,
  })

  // absolute path
  await createSymlinks('*', {
    cwd,
    dest: '/etc/nginx/sites-available',
    dry: true,
  })

  await createSymlinks(path.resolve(process.cwd(), cwd), {
    dest: '/etc/nginx/sites-enabled',
    dry: true,
  })
})

it('should created symlinks', async () => {
  const { globby } = await import('globby')
  const paths = await globby('*', { absolute: true, cwd: dest })

  for (const item of paths) {
    const link = await fs.readlink(item)

    expect(link).toBeTypeOf('string')
  }
})
