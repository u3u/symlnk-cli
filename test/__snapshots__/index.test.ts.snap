// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshots > cancel clean 1`] = `
[
  "D temp/symlinks/a",
  "D temp/symlinks/b",
  "ℹ Canceled.",
]
`;

exports[`should match snapshots > cancel create 1`] = `
[
  "M test/a -> __fixtures__/a",
  "M test/b -> __fixtures__/b",
  "ℹ Canceled.",
]
`;

exports[`should match snapshots > clean 1`] = `
[
  "✔ temp/symlinks/a",
  "✔ temp/symlinks/b",
]
`;

exports[`should match snapshots > create 1`] = `
[
  "M temp/symlinks/a -> ../../test/__fixtures__/a",
  "M temp/symlinks/b -> ../../test/__fixtures__/b",
  "✔ temp/symlinks/a -> ../../test/__fixtures__/a",
  "✔ temp/symlinks/b -> ../../test/__fixtures__/b",
]
`;

exports[`should match snapshots > no match 1`] = `
[
  "ℹ No matching files found.",
]
`;

exports[`should match snapshots > skip exists 1`] = `
[
  "skip temp/symlinks/a -> ../../test/__fixtures__/a",
  "skip temp/symlinks/b -> ../../test/__fixtures__/b",
]
`;
